Add Clone Flow to Jetpack Cloud

jwebbdev opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

## Proposed Changes

* Implements the clone flow into Jetpack Cloud

- [x] Implement Set Destination Screen
- [x] Implement Select Clone Period Screen
- [x] Implement Configure Screen
- [x] Detect Clone Progress
- [x] Implement Clone Progress Screen
### Known Issues
 * Switching from the Clone flow to Settings retains the state of 'updateRequestStatus' so can show the Finished screen.  Clicking the Finish Up button will return to normal with no ill effects.  The solution here is likely to have the state track a combination of siteId and role for these operations.  Since this isn't breaking and an edge case we're going to follow-up with a fix instead of making this PR larger.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

To test cloning you'll need two sites.  A source and a destination.  

The source site needs to have a Jetpack Backup plan and at least one backup.
The destination site needs FTP or SSH/SFTP access, and you'll need the credentials and server info handy.
You can utilize Jurassic Ninja to create both sites if needed.
You may want to make some changes so it'll be obvious the clone took place afterwards.

* Open the Jetpack Cloud Live link
* Select your source site
* Visit the VaultPress backup area
* Add `/clone` to the url - This is temporary until we add links
* Enter your server credentials and test them.  Once they're valid, click 'Finish Up'
* Select the Backup point you wish to clone
* Select the types of items, or leave the default of everything and Confirm
* Wait for the restore to finish and visit the destination site.

### UI Changes to existing pages
* Backup main page has had a header added with a Clone button
* Activity log main page has had the header changed, and includes a Clone button

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

## Screenshots for Translation Context
<img width="771" alt="Screenshot 2023-04-03 at 10 02 37 AM" src="https://user-images.githubusercontent.com/1992658/229534378-bbb1b968-7d2f-44ac-9659-013bf3dd2b47.png">
<img width="750" alt="Screenshot 2023-04-03 at 10 02 46 AM" src="https://user-images.githubusercontent.com/1992658/229534456-fb0eee3b-75e4-4a2b-9e43-d100656501c2.png">
<img width="726" alt="Screenshot 2023-04-03 at 10 05 53 AM" src="https://user-images.githubusercontent.com/1992658/229534474-bbe12d7d-3833-4d76-a67c-bc09fcdec020.png">
<img width="750" alt="Screenshot 2023-04-03 at 10 03 23 AM" src="https://user-images.githubusercontent.com/1992658/229534556-2dbe71e0-88ab-407c-8a4c-c7d8668fdbe8.png">
<img width="770" alt="Screenshot 2023-04-03 at 10 07 01 AM" src="https://user-images.githubusercontent.com/1992658/229534702-68b42d59-0596-4dd2-8a58-418d6711126b.png">
<img width="895" alt="Screenshot 2023-04-03 at 10 03 37 AM" src="https://user-images.githubusercontent.com/1992658/229535072-e07a8571-384e-41ba-a3e4-bb35b89ad2fc.png">
<img width="774" alt="Screenshot 2023-04-03 at 10 03 43 AM" src="https://user-images.githubusercontent.com/1992658/229534742-6a740ca3-9d62-43b3-86c3-f8280ba9861f.png">
<img width="787" alt="Screenshot 2023-04-03 at 10 05 29 AM" src="https://user-images.githubusercontent.com/1992658/229534779-fe1549e8-b306-4c8b-97e3-67c47e2e9f50.png">

jwebbdev committed:
Initial scaffolding for clone flow in cloud

jwebbdev committed:
Added better wrapping styles / components

jwebbdev committed:
Update credential form to handle role and site_url

jwebbdev committed:
All onFinishUp callback for verification to support cloning

jwebbdev committed:
Added temporary period selection for testing

jwebbdev committed:
Fix linter issues

jwebbdev committed:
Fix clone progress tracking

jwebbdev committed:
Fixed linting

mavegaf committed:
Add/backup clone select point (#74726)

* Show rewind items in clone section

* Adding props 'availableActions' to decide which option to show (clone vs restore/download)

* Connect clone action with method onSetBackupPeriod

jwebbdev committed:
Style fixes

jwebbdev committed:
Merge remote-tracking branch 'origin/add/backup-clone-flow' into add/backup-clone-flow

jwebbdev committed:
Merge remote-tracking branch 'origin/trunk' into add/backup-clone-flow

jwebbdev committed:
Add more specific class declaration to fix out of order css issues

jwebbdev committed:
Add clone-only info text for credentials

jwebbdev committed:
Add optional back button for alternate creds

jwebbdev committed:
Change labels for clone flow

jwebbdev committed:
Add clone button to backup main page

jwebbdev committed:
Activity Log header with clone button

jwebbdev committed:
Linting fixes

jwebbdev committed:
Remove temporary button

ederrengifo commented:
Few changes to apply based on previous coordination:

- [ ] Change the action from "cloning" to "copying" (double check the latest design's copy)
- [ ] The call to action must be a secondary button style
- [ ] Both call to action in Cloud includes a tooltip when mouseover (check designs)
- [ ] Update the navigation bar from the one we already use to get a new license (for partners)
- [ ] Standardize the use and positioning of secondary and primary buttons for the steps (aligned to the left)
- [ ] Update "point to copy" step (WIP)
- [ ] Update views to check the progress of the copy and success scenario (check designs)

jwebbdev committed:
Merge remote-tracking branch 'origin/trunk' into add/backup-clone-flow

jwebbdev committed:
Rename Clone actions to Copy

jwebbdev committed:
Change copy CTA to secondary button style

jwebbdev committed:
Add tooltip to Copy this site buttons

jwebbdev committed:
Update back buttons in clone flow

jwebbdev committed:
Linter fixes

jwebbdev committed:
Change clone to copy in text

mavegaf committed:
Adding tests to BackupCloneFlow component. (#74844)

* Adding tests to BackupCloneFlow component.
** RewindFlowLoading is rendered if state is not initialized
** AdvancedCredentials is rendered if user has not set the destination
** ActivityCardList is rendered if user has not set the period to clone
** RewindConfigEditor is rendered if user has not requested yet the restore
** Render ProgressBar if uses requested the clone
** Render ProgressBar if inProgressRewindStatus is queued
** Render ProgressBar if inProgressRewindStatus is running
** Render finished text if the clone is finished
** Render error if it is in a wrong status
* Test action-buttons depending if available action is clone or restore/download

jwebbdev committed:
Add configuration confirmation screen

jwebbdev committed:
Allow RewindStatus to return status with main creds

jwebbdev committed:
Merge branch 'add/backup-clone-flow' of github.com:Automattic/wp-calypso into add/backup-clone-flow

jwebbdev committed:
Added type info

jwebbdev committed:
Track and display alternate credential url

jwebbdev committed:
Update styles for Clone Flow progress

jwebbdev committed:
Updates to configuration screen

jwebbdev committed:
Fix gridicons-external button style

mavegaf committed:
Add/highlight last backup clone point (#75051)

* Highlight the last backup point when selecting a backup to clone

jwebbdev committed:
Merge remote-tracking branch 'origin/trunk' into add/backup-clone-flow

jwebbdev committed:
Implement License style StepProgress

jwebbdev committed:
Removed completed TODOs

Initsogar commented on client/my-sites/activity/activity-log-v2/index.tsx:
```diff
@@ -39,12 +42,28 @@ const ActivityLogV2: FunctionComponent = () => {
+			</div>
+			<div className="activity-log-v2__header-right">
+				{ selectedSiteSlug && (
+					<Tooltip
+						text={ translate(
+							'To test your site changes, migrate or keep your data safe in another site'
+						) }
+					>
+						<Button
+							className="activity-log-v2__clone-button"
+							href={ backupClonePath( selectedSiteSlug ) }
+						>
+							{ translate( 'Copy this site' ) }
+						</Button>
+					</Tooltip>
```
Not a blocker, but maybe we could consider making this reusable, as it is also required on the backup page and just differs on `className`.

Initsogar commented on client/my-sites/backup/clone-flow/index.tsx:
```diff
@@ -0,0 +1,448 @@
+import './style.scss';
+
+interface Props {
+	siteId: number;
```
Should `siteId` be a required prop? It appears that the majority of selectors and actions depend on it.

Initsogar commented on client/my-sites/backup/clone-flow/index.tsx:
```diff
@@ -0,0 +1,448 @@
+		[ dispatch, siteId ]
+	);
+	const backupCurrentlyInProgress = useSelector( ( state ) => {
+		if ( ! siteId ) {
```
Given that `siteId` is a required prop, is this validation necessary?

Initsogar commented on client/my-sites/backup/clone-flow/index.tsx:
```diff
@@ -0,0 +1,448 @@
+	const renderSetBackupPeriod = () => (
+		<>
+			<CloneFlowStepProgress currentStep="clonePoint" />
+			<h3 className="clone-flow__title">{ translate( 'Select a Backup Point to Copy' ) }</h3>
```
Maybe we could use caps only for the first word.

```suggestion
			<h3 className="clone-flow__title">{ translate( 'Select a backup point to copy' ) }</h3>
```

jwebbdev committed:
Update tests to use CloneFlowStepProgress

Initsogar commented on client/my-sites/backup/clone-flow/test/index.jsx:
```diff
@@ -0,0 +1,468 @@
+}
+
+function getLogs() {
+	return {
```
Maybe we could consider moving the fixtures into a separate files like [this example](https://github.com/Automattic/wp-calypso/blob/20fc989ed71ca1b6202043ed0c2a71ace965ed8e/client/reader/discover/test/fixtures/index.js) to make this file smaller.

Initsogar reviewed (approved):
Great work team!

Tested the clone flow using Jurassic Ninja sites from Activity Log and Backup page and works great.

Just left a few comments that are not stoppers, so you could decide to address them in this PR or in a next iteration.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7895803

Hi @jwebbdev, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Enter the server credentials from your hosting provider. {{a}}Learn how to find and enter your credentials{{/a}}`
- `Please enter a valid destination site url.`
- `Confirm credentials`
- `Clone from latest point`
- `This is the complete event history for your site`
- `To test your site changes, migrate or keep your data safe in another site`
- `Copy this site`
- `Set a destination site`
- `Input information about the site you want to clone to`
- `Select a Backup Point to Copy`
- `Which point in your site's history would you like to copy from?`
- `Before continue, be aware that any current content on {{strong}}%(destinationUrl)s{{/strong}} will be overriden based on what you configured to copy.`
- `Select the items you want to copy to {{strong}}%(destinationUrl)s{{/strong}}.`
- `{{strong}}%(backupDisplayDate)s{{/strong}} is the selected point to copy.`
- `Copying will override and remove all content on the destination site.`
- `A backup is currently in progress; copying now will stop the backup.`
- `Confirm configuration`
- `Copying site to %(destinationUrl)s`
- `Initializing the copy process`
- `Jetpack is copying your site. You will be notified when the process is finished in the activity log.`
- `Go to Activity Log`
- `Your site has been successfully copied.`
- `All of your selected items are now copied from the {{strong}}%(backupDisplayDate)s{{/strong}} backup.`
- `Copy failed: %s`
- `An error occurred while restoring your site. Please {{button}}try your copy again{{/button}} or contact our support team to resolve the issue.`
- `Set destination`
- `Select clone point`
- `Latest Backups`
- `This is a list of your latest generated backups`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7895803).
